pallets/liquid-staking-two/src/lib.rs

```diff
@@ -0,0 +1,820 @@
+                        claim_unstake_each_era = user_buffer
+                            .total_unstake_amount
+                            .saturating_sub(user_buffer.claimed_unstake_amount);
+
+                        claim_stake_each_era = user_buffer
+                            .total_stake_amount
+                            .saturating_sub(user_buffer.claimed_stake_amount);
+                    }
+                    _ => (),
+                };
+
+                MatchingQueue::<T>::try_mutate(who, &era_index, |b| -> DispatchResult {
+                    b.claimed_unstake_amount = b
+                        .claimed_unstake_amount
+                        .saturating_add(claim_unstake_each_era);
+                    b.claimed_stake_amount =
+                        b.claimed_stake_amount.saturating_add(claim_stake_each_era);
+
+                    if b.total_stake_amount == b.claimed_stake_amount
+                        && b.total_unstake_amount == b.claimed_unstake_amount
+                    {
+                        // user have already claimed all he can claim in this era, remove it from MatchingQueue
+                        remove.push(era_index);
+                    }
+                    Ok(())
+                })
+                .ok()
+                .and_then(|_| {
+                    withdrawable_unstake_amount =
+                        withdrawable_unstake_amount.saturating_add(claim_unstake_each_era);
+                    withdrawable_stake_amount =
+                        withdrawable_stake_amount.saturating_add(claim_stake_each_era);
+                    Some(())
+                })
+            });
+        });
+
+        // remove finished records from MatchingQueue
+        if remove.len() > 0 {
+            remove.iter().for_each(|era_index| {
+                MatchingQueue::<T>::remove(who, era_index);
+            });
+        }
+
+        // transfer xKSM from palletId to who
+        if withdrawable_stake_amount > 0 {
+            let xtoken_amount = ExchangeRate::<T>::get()
+                .reciprocal()
+                .and_then(|r| r.checked_mul_int(withdrawable_stake_amount))
+                .ok_or(Error::<T>::InvalidExchangeRate)?;
+            T::Currency::transfer(
+                T::LiquidCurrency::get(),
+                &Self::account_id(),
+                who,
+                xtoken_amount,
+            )?;
+        }
+
+        // transfer KSM from palletId to who
+        if withdrawable_unstake_amount > 0 {
+            T::Currency::transfer(
+                T::StakingCurrency::get(),
+                &Self::account_id(),
+                who,
+                withdrawable_unstake_amount,
+            )?;
+        }
+
+        Ok(().into())
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    pub fn account_id() -> T::AccountId {
+        T::PalletId::get().into_account()
+    }
+}
+
+impl<T: Config> ExchangeRateProvider for Pallet<T> {
```

Comment:
This can be easily added if need.

Reply:
Removed

Reply:
we absolutely need it, why remove it?
https://github.com/parallel-finance/parallel/blob/master/runtime/heiko/src/lib.rs#L805